source/map.js

```diff
@@ -45,6 +45,8 @@ import keys from './keys';
 var map = _curry2(_dispatchable(['fantasy-land/map', 'map'], _xmap, function map(fn, functor) {
   switch (Object.prototype.toString.call(functor)) {
     case '[object Function]':
+    case '[object AsyncFunction]':
+    case '[object GeneratorFunction]':
```

Comment:
```js
case '[object AsyncGeneratorFunction]':
```

Reply:
Have added AsyncGeneratorFunction as an extra case, thanks for pointing this out!

Reply:
As this currently stands it is not a valid functor implementation because the provided `fn` is forced to receive a Generator of `a`, or Promise of `a` rather than accepting functions that work for any `a`.

I.e. it should be possible to do something like `R.map(a => a + 1, function*(a) { yield a * 2 })` 

By passing in the generator or promise, you’re providing more flexibility or power than functors alone should have access to.

We could however update it with their own seperate implementations with something like:
```
case '[object AsyncFunction]':
return async a => fn(await functor(a))

case '[object GeneratorFunction]':
return function*() {
  for (let a of functor()) yield fn(a);
}
```
The only downside of this is that it then makes JS runtimes with async and generator support a requirement to use Ramda, even if you don’t plan on using the functionality.

Reply:
Yes, of course.  I was looking at the entirely wrong bit of rules for problems.

I like your alternative... but probably only for v1.0+

Reply:
This is interesting, however I wonder if treating AsyncFunctions (typeof fn === 'AsyncFunction') as separate from functions that return Promises (typeof fn === 'Function') could have unintended consequences.

It would mean babel can change your program's behaviour (e.g. if you start or stop transpiling, `map` will treat your transpiled [Special]Functions differently!), but also it breaks eta abstraction/reduction:
```
const f = tap(console.log)
const asyncFunction = async x => x

map(f, async x => x)(42) //"42"
map(f, x => asyncFunction(x))(42) //Promise<42>
//because
Object.prototype.toString.call(asyncFunction) //'[object AsyncFunction]'
Object.prototype.toString.call(x => asyncFunction(x)) //'[object Function]'
```
(This also extends to `compose`, `pipe` etc.)

In a typeless language like JS where we really can't know what a function will return and there's no way at runtime time to truly know if `x => f(x)` really is the same as `f` then I feel like we should rely on the extensional equality of values rather than the intensional properties that `Object.prototype.toString.call` sometimes exposes, but other times can't.

At the moment, Ramda already treats AsyncFunctions as sometimes separate entities.
In the transpiled ramda repl:
```
const asyncFn = async x => x
map(tap(console.log), asyncFn)(5) //[object Promise]
```
but in an ES6 native environment like node:
```
const asyncFn = async x => x
map(tap(console.log), asyncFn) //undefined
map(tap(console.log), asyncFn)(5)
//TypeError: R.map(...) is not a function
map(tap(console.log), x => asyncFn(x))(5)
//Promise { 5 }
map(tap(console.log), compose(asyncFn))(5)
//Promise { 5 }
```

My PR as it stands intentionally treats AsyncFunctions and GeneratorFunctions as just normal function functors, which is transpilation and eta-abstraction proof. Same goes for any other functor that has a consistent perceived type regardless of abstraction - e.g. arrays, objects or something that has a delegateable map e.g. eithers and maybes.

If we do need to treat AsyncFunctions and GeneratorFunctions (or, abstracted, functions that return Promises and functions that return Generators) differently when mapping, it seems safer to have a separate `map` (or `mapWith`...) that treats them consistently regardless of the environment.